<a href="https://colab.research.google.com/github/khusanovich/Machine_Learning/blob/main/05_ML_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![Imgur](https://i.imgur.com/5pXzCIu.png)

# Data Science va Sun'iy Intellekt Praktikum

## 5-MODUL. Machine Learning

### 5.1 - ML loyiha qadamlari




Esingizda bo'lsa, biz eng birinchi darsimizni Data Science Metodologiyasini o'rganishdan boshlagan edik. ML ham Data Science loyihaning bir qismi bo'lgani sabab, har qanday loyihani bajarishda biz avval ko'rgan metodologiyaga amal qilamiz.

<img src="https://i.imgur.com/dzZnnYi.png" alt="CRISP-DM" width="800"/>


Machine Learning qanday ishlashini tushunish uchun ushbu darsimiz davomida biz bitta loyihani boshidan oxirigacha birgalikda qilamiz. 

Dars davomida tushunmagan mavzularga duch kelsangiz, havotir qilmang, kurs davomida juda ko'p masalalarga qayta to'xtalamiz.

### 1-QADAM. FAOLIYATNI O'RGANISH

Keling boshlanishiga loyiha haqida tasavvur hosil qilaylik.

Sizning buyurtmachingiz "California Inc" Qurilish Kompaniyasi. Kompaniya keyingi uylarini qurishda California shtatining turli hududlarida uylarning o'rtacha narxini hisoblab beruvchi model (dastur) yaratishni so'rayapti.

Sizga berilgan ma'lumotlar: California shtatida aholini ro'yxatga olish uchun o'tkazilgan so'rovnoma natijasi.

Ma'lumotlar Californianing har bir hududi (__blok__ deb ataladi) uchun aholi soni, aholining median daromadi, uylarning median narxi kabi qiymatlarni o'z ichiga oladi. Har bir hududda tahminan 600 dan 3000 gacha aholi yashaydi.

**Demak sizning vazifangiz yuqorida ma'lumotlarga asoslangan holda turli hududlarda uylarning median narxini aniqlab beruvchi model (dastur) yaratish**

#### 1.1 Jarayonlarni tushunish

Sizning birinchi qiladigan ishingiz, loyiha haqida ko'proq ma'lumot to'plash. Bu sizga loyihaga to'g'ri yondoshish, to'g'ri model tanlash, qanday algoritmlardan foydalanishni aniqlash uchun yordam beradi. 

Misol uchun, birinchi savol: Ushbu model qayerda ishlatiladi?
Buyurtmachi sizning modelingdan qaytgan natija boshqa, siz bilmaydigan, ma'lumotlar bilan birgalikda yana boshqa modelning kirishiga ulanishini va yakuniy model ko'rsatilgan hududga investisiya kiritish yoki kiritmaslik haqida qaror qabul qilishini aytdi. 

Demak, sizning modelingiz kattaroq modelga ulanar ekan. 

Keyingi savol: Hozirda buyurtmachi median narxni aniqlashda qanday usuldan foydalanayapti va joriy usulning aniqligi nechiga teng, uning kamchiliklari nimada?

Javob: Hozirda hudud uchun narxlar bir nechta mutaxassislarning tahliliga asosan, qo'lda aniqlanmoqda. Aniqligi 20%. 

Mana endi keyingi bosqichga o'tamiz.

#### 1.2 Analitik yondoshuvni aniqlash
Bu bosqichda biz bir nechta narsalarni aniqlashtirib olamiz: 
- Model: Supervised, Unsupervised, Reinforcement?
- Algoritm: klassifikasiya, regressiya, yoki boshqa turda
- Usul: Online yoki offline

Yuqoridagi savollarga javob beramiz:
1. Supervised learning - sababi bizda ma'lumotlarda label (yorliq) mavjud. Bu hududdadi median narx. Biz ham aynan shu narxni bashorat qilmoqchimiz.
2. Regressiya - "Bashorat" (prognoz) dedikmi demak bu regressia algoritmlari yordamida hal qilinadi
3. Offline - sababi ma'lumotlar bizga avvaldan bir marta berilgan. Doimiy ma'lumotlar oqimi yo'q.

Model aniqligini qanday baholaymiz?

Aniqlikni baholashning turli usullari bor, regressiya algoritmlar uchun odatda **o'rtacha kvadrat xatolik** (Root Mean Square Error - RMSE) ko'p ishlatiladi:

![RMSE](https://i.imgur.com/aXL9iWa.png)

Bu yerda:
- $m$ - datasetdagi qatorlar soni (har bir qator bitta ma'lumot)
- $x^{(i)}$ - $i$-qator uchun barcha parametrlar vektori (_label_ dan tashqari)
- $y^{(i)}$ - $i$-qator uchun label (bizdagi misolda median uy narxi)
- $X$ - labeldan boshqa barcha parametrlar
- $h$ - sizning modelingizdan qaytgan bashorat (hypothesis). 
    - $h(x^{(i)})$ - $i$-qator uchun model qaytargan bashorat.
    
Aniqlikni baholashning yana bir usuli, o'rtacha absolyut xatolik (mean absolute error - MAE).

![MAE](https://i.imgur.com/GJH9CGy.png)

RMSE ham MAE ham ikki vektor, bashorat va label o'rtasidagi farqni hisoblaydi. Xato qancha kam bo'lsa, natija shuncha yaxshi hisoblanadi.

Keyingi bosqichga o'tishdan avval, yuqoridagi xulosalarimizni tasdiqlab olamiz. Sababi, sizning regressiya modelingiz aniq son qaytaradi (uyning bashorat qilingan narxi), lekin buyurtmachi sizdan son emas "qimmat", "arzon", "o'rtacha" kabi qiymatlarni kutgan edi. Bu esa klassifikasiya algoritmi yordamida hal qilinishi kerak edi. Demak, siz modelingizdan qaytadigan natija buyurtmachi kutgan natija bilan bir ekanini aniqlashtirib olamiz.

### 2-QADAM. MA'LUMOTLARNI O'RGANAMIZ

#### !DIQQAT. Keyingi qadamlarni bajarishdan avval, loyiha uchun yangi muhit yaratib olish tavsiya qilinadi. Umuman olganda har bir loyiha uchun alohida muhit yaratish yaxshi odat!


<img src="https://i.imgur.com/YcFHBRK.png" alt="Anaconda" width="800"/>

Kerakli modullarni chaqirib olamiz. Bu safar bizga avvaldan tanish `pandas`, `numpy` kutubxonalaridan tashqari, yangi, `scikit-learn` kutubxonaisini ham chaqiramiz. 

**[scikit-learn](https://scikit-learn.org/stable/)** - turli Machine Learning algoritmlarni jamlagan kutubxona bo'lib, kelgusida ishimizni bir necha bor osonlashtiradi. 

In [ ]:
import pandas as pd
import numpy as np
import sklearn # scikit kutubxonasi
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
URL = "http://github.com/ageron/handson-ml2/blob/master/datasets/housing/housing.csv?raw=true"
df = pd.read_csv(URL)

In [ ]:
df.head()

longitude  latitude  ...  median_house_value  ocean_proximity
0    -122.23     37.88  ...            452600.0         NEAR BAY
1    -122.22     37.86  ...            358500.0         NEAR BAY
2    -122.24     37.85  ...            352100.0         NEAR BAY
3    -122.25     37.85  ...            341300.0         NEAR BAY
4    -122.25     37.85  ...            342200.0         NEAR BAY

[5 rows x 10 columns]

Datasetning har bir qatori bu bitta hudud (mahalla, kvartal, blok) haqida ma'lumot. Dataset 10 ta ustundan iborat:
- `longitude` va `latitude` birgalikda GPS koordinatalar
- `housing_median_age` - uylarning median yoshi
- `total_rooms` - ushbu hududda mavjud jami xonalar
- `total_bedrooms` - ushbu hududda mavjud jami yotoqxonalar
- `population` - hudud aholi soni
- `households` - hududdagi oilalar soni
- `median_income` - median daromad
- `median_house_values` - uylarning median narxi
- `ocean_proximity` - hududning okeanga qanchalik yaqinligi

### ❓ Median va o'rtacha qiymatlar o'rtasida farqni bilasizmi? 

Deylik bizda 10 ta son bor, shu sonlarning o'rtachasini hisoblash uchun avval ularni qo'shamiz keyin esa yig'indini 10 ga bo'almiz.
Median hisoblash uchun esa sonlarni o'sish tartibida tahlaymiz va o'rtadagi sonni olamiz.

![Imgur](https://i.imgur.com/Qkzwv3u.png)

### 2.1 Ma'lumotlarni ko'ramiz qilamiz

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


## ocean proximity bizda matn turidagi ma'lumotlarni saqlagan

In [ ]:
df.head()

longitude  latitude  ...  median_house_value  ocean_proximity
0    -122.23     37.88  ...            452600.0         NEAR BAY
1    -122.22     37.86  ...            358500.0         NEAR BAY
2    -122.24     37.85  ...            352100.0         NEAR BAY
3    -122.25     37.85  ...            341300.0         NEAR BAY
4    -122.25     37.85  ...            342200.0         NEAR BAY

[5 rows x 10 columns]

Yuqoridagi natijadan olingan xulosalar:
1. Dataset 20640 qatordan iborat (har bir qator bu alohida hudud (blok, mahalla, kvartal))
2. `total_bedrooms` ustunida ma'lumotlar to'liq emas (20433 qator to'liq, qolgan 207 tasida ma'lumot yo'q)
3. `ocean_proximity` ustunidan boshqa barcha ustunlar sonli.

`ocean_proximity` ustunini ko'ramiz:

In [ ]:
df['ocean_proximity'].value_counts()

<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: ocean_proximity, dtype: int64

Demak bu ustunda uylarning okeandan qanchalik uzoqligi matn ko'rinishida saqlangan. Har bir matnni tarjima qilishni sizga vazifa qilib qoldiramiz. 
Biz esa ma'lumotlarni tahlil qilishda davom etamiz.

In [ ]:
df.describe()

longitude      latitude  ...  median_income  median_house_value
count  20640.000000  20640.000000  ...   20640.000000        20640.000000
mean    -119.569704     35.631861  ...       3.870671       206855.816909
std        2.003532      2.135952  ...       1.899822       115395.615874
min     -124.350000     32.540000  ...       0.499900        14999.000000
25%     -121.800000     33.930000  ...       2.563400       119600.000000
50%     -118.490000     34.260000  ...       3.534800       179700.000000
75%     -118.010000     37.710000  ...       4.743250       264725.000000
max     -114.310000     41.950000  ...      15.000100       500001.000000

[8 rows x 9 columns]

Yuqoridagi jadvladan qanday xulosalar oldingiz?

Ma'lumotlarni tahlil qilishning yana bir usuli - vizualizasiya